Fix Valhalla test compilation

theresa-m opened the pull request:
- update Valhalla tests to use NullRestricted annotations instead of primitive
- don't allow @NullRestricted annotation for arrays or primitive types

this won't compile without https://github.com/eclipse-openj9/openj9/pull/19631#pullrequestreview-2099915816

Fixes: https://github.com/eclipse-openj9/openj9/issues/19459

theresa-m commented:
I rebased with the changes from https://github.com/eclipse-openj9/openj9/pull/19631. Compiling the tests should pass now.

hangshao0 commented on test/functional/Valhalla/src_qtypes/org/openj9/test/lworld/NullRestrictedTypeOptTests.java:
```diff
@@ -69,46 +82,60 @@ static public void testEAOpportunity1() throws Throwable {
 	}
 
+	@NullRestricted
+	static PrimPair evalTestEA1a_p1;
```
Is there a reason to change local variables p1 to p4 to static variables ? Defining them as static might change what these `evalTestEAXX` tests are originally designed for (exercising escape analysis).

theresa-m replied:
The annotation is not defined for local variables.

hangshao0 replied:
If the test case is not testing null, whether the variable is null restricted or not does not matter. We can just use local variable like `PrimPair p1`

a7ehuo replied:
> Defining them as static might change what these evalTestEAXX tests are originally designed for (exercising escape analysis). 

I have the same concern. If they are not local variables, EA might not be triggered. With the latest spec change, I'm not too sure if all these `evalTestEA* `tests would still behave as expected.  I'll let @hzongaro comment further

hzongaro replied:
I agree with @hangshao0 and @a7ehuo here - the point of the test was to exercise Escape Analysis with null-restricted types.  By assigning the value to a static field, it automatically escapes, so there's no longer an opportunity for stack allocation.

I would suggest simply disabling the test for now, just in case null-restricted types ever become first class types.

hzongaro replied:
Sorry - let me withdraw that suggestion.  I think it would be OK to keep the test active, and instead change the static fields back to local variables.  Even if the local variables cannot be null-restricted, there are still opportunities for Escape Analysis to perform stack allocation.

theresa-m replied:
Thanks Henry, I will update the tests with your suggestion. Once javac is updated and we transition to the tests in src_lw5 the fields can be marked as null restricted https://github.com/eclipse-openj9/openj9/blob/master/test/functional/Valhalla/src_lw5/org/openj9/test/lworld/NullRestrictedTypeOptTests.java

hzongaro replied:
Sorry.  When I was reviewing this, I didn't realize that this was for the src_qtypes version of the test rather than the src_lw5 version of the test.  I think your updates look good.

a7ehuo replied:
LGTM. Thanks for the update!

hangshao0 commented on test/functional/Valhalla/src_qtypes/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -3029,7 +3026,6 @@ static private void valueTypeIdentityObjectTestHelper(String testName, String su
 			superClassName = "java/lang/Object";
 		}
 		Class valueClass = ValueTypeGenerator.generateValueClass(testName, superClassName, fields, extraClassFlags);
-		assertTrue(PrimitiveClass.isPrimitiveClass(valueClass));
```
Instead of removing the assertion, can it be replaced with `assertTrue(valueClass.isValue());`

hangshao0 commented on test/functional/Valhalla/src_qtypes/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -3100,50 +3096,10 @@ static public void testIdentityObjectOnRef() throws Throwable {
-	static public void testIsPrimitiveClassOnRef() throws Throwable {
-		String fields[] = {"longField:J"};
-		Class refClass = ValueTypeGenerator.generateRefClass("testIsPrimitiveClassOnRef", fields);
-		assertFalse(PrimitiveClass.isPrimitiveClass(refClass));
```
Can `PrimitiveClass.isPrimitiveClass(refClass)` be replaced with `refClass.isValue()` ? 
The same comment applies to the` PrimitiveClass.isPrimitiveClass()` cases removed below.

theresa-m replied:
That scenario is covered by testIdentityObjectOnRef.

hangshao0 replied:
Do we have test cases for `isValue()` and `isIdentity()` on interface, value array and ref arrays ?

hangshao0 replied:
Do we have test cases for isValue() and isIdentity() on interface, value array and ref arrays ?

theresa-m replied:
Looks like there is not. I can modify these tests to use isValue instead.

theresa-m replied:
I added these tests.

hangshao0 commented on test/functional/Valhalla/src_qtypes/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -3180,26 +3136,6 @@ static public void testValueClassHashCode() throws Throwable {
 	}
 
-	@Test(priority=1)
-	static public void testPrimitiveClassHashCode() throws Throwable {
```
Is there a reason why this case is removed ?

theresa-m replied:
This to match what was removed in the lw5 test suite https://github.com/eclipse-openj9/openj9/pull/18505#issue-2004727285

hangshao0 replied:
I see. There is `testValueClassHashCode()` above this.

hangshao0 commented:
@a7ehuo @hzongaro `NullRestrictedTypeOptTests.java` was added/reviewed by you. I am including you as reviewers here.

a7ehuo commented on test/functional/Valhalla/src_qtypes/org/openj9/test/lworld/NullRestrictedTypeOptTests.java:
```diff
@@ -52,7 +58,14 @@ public EscapeException(Object o) {
+	static {
+		parr = (PrimPair[])ValueClass.newArrayInstance(
+			NullRestrictedCheckedType.of(PrimPair.class), 3);
+		parr[0] =  new PrimPair(3, 4);
```
Very minor cosmetic comment: one extra space before `new`
```c++
		parr[0] =  new PrimPair(3, 4);
```
=>
```c++
		parr[0] = new PrimPair(3, 4);
```

a7ehuo reviewed (commented):
Thank you very much for making this change @theresa-m!  Very minor comments for `NullRestrictedTypeOptTests.java`

theresa-m commented:
Thanks @hangshao0 and @a7ehuo for your reviews. I have addressed all comments except the conversation of moving variables outside methods. The second force push was to restart the line ending check.

hzongaro commented on test/functional/Valhalla/src_qtypes/org/openj9/test/lworld/NullRestrictedTypeOptTests.java:
```diff
@@ -69,46 +82,60 @@ static public void testEAOpportunity1() throws Throwable {
+		result += evalTestEA1a_p3.x + evalTestEA1a_p4.y;
 	}
 
+	@NullRestricted
+	static PrimPair evalTestEA1b_p1;
+	@NullRestricted
+	static PrimPair evalTestEA1b_p2;
+
+	@NullRestricted
+	static PrimPair evalTestEA1b_p3;
+	@NullRestricted
+	static PrimPair evalTestEA1b_p4;
+
```
As with test `evalTestEA1a` above, making these previously local variables into static fields prevents Escape Analysis from stack allocating.  I would suggest disabling `evalTestEA1b` for now.

hzongaro replied:
As I suggested above, maybe the test can be kept active, with the static fields reverted to local variables.  There are still opportunities for Escape Analysis to perform stack allocation, despite the fact that the variables cannot be marked as null-restricted.

hangshao0 commented on test/functional/Valhalla/src_qtypes/org/openj9/test/lworld/ValueTypeUnsafeTests.java:
```diff
@@ -307,65 +322,87 @@ static public void testNonVTClzGetValue() throws Throwable {
 	}
 
+	@NullRestricted
+	static ZeroSizeValueTypeWrapper zsvtw;
```
We should keep this as local variable, and put the `Unsafe.getValue()` inside `if (Unsafe.isFieldAtOffsetFlattened()) { }`. 

Similarly for the `Unsafe.getValue(), Unsafte.putValue()` test cases in this file, they can be kept as local variables. We just need to add additional check like `Unsafe.isFlatArray()`,` Unsafe.isFlatField()` or `Unsafe.isFieldAtOffsetFlattened()` before them.

hangshao0 replied:
Discussed on slack, we can keep these variables as locals (without changing them to null restricted static) if the tests are already passing.

hangshao0 commented on test/functional/Valhalla/src_qtypes/org/openj9/test/lworld/ValueTypeUnsafeTests.java:
```diff
@@ -245,13 +253,18 @@ static public void testValueHeaderSizeOfArray() throws Throwable {
 		assertEquals(size, 0);
 	}
 
+	@NullRestricted
```
> It is fine to let this test fail now (track the failure in an issue)

I was assuming the test failed without this change. If the test is passing (with these 2 variables being local), there is no need to track.

theresa-m commented:
After discussing with Hang we concluded the Unsafe tests will need a closer look and likely modifications to the class library. I will make a separate issue for this.

hangshao0 commented:
Any more review comments ? @hzongaro @a7ehuo

hzongaro reviewed (approved):
Looks good.  Thanks!

a7ehuo commented:
LGTM. Thanks for the update!

hangshao0 commented:
Jenkins test sanity,extended zlinuxval jdknext

hangshao0 commented:
Merging this as this PR is to fix the test compilation, which passed in the PR build. The test failures will be addressed in future PRs.
